File: js/complete/no008.js
#!/usr/local/bin/node

/* Find the greatest product of five consecutive
   digits in the 1000-digit number. */

var fns = require('../functions.js'),
    operator = require('../operator.js'),
    timer = require('../timer.js');

function productConsecDigits(number, consecutive) {
    /* Returns the largest product of "consecutive"
       consecutive digits from number */
    var digits = number.toString().split(''),
        max_start = digits.length - consecutive,
        result = [];

    /* Could track the product instead of reducing each time.
       Just doing what python does. */
    for (var i = 0; i < max_start + 1; i++) {
	result.push(digits.slice(i, i + consecutive).reduce(operator.mul, 1));
    }

    return result;
};

function main() {
    var data = fns.get_data(8).split('\n').join(''),
        products = productConsecDigits(data, 5);
    return Math.max.apply(Math, products);
};

timer.timer(8, main);

File: python/complete/no008.py
#!/usr/bin/env python

# Find the greatest product of five consecutive
# digits in the 1000-digit number.

from python.decorators import euler_timer
from python.functions import get_data


def my_reduce(iterable):
    acc = 1
    for val in iterable:
        acc = acc * val
    return acc


def my_max(iterable):
    maximum = 0
    for val in iterable:
        if val > maximum:
            maximum = val
    return maximum


def product_consec_digits(number, consecutive):
    """
    Returns the largest product of "consecutive"
    consecutive digits from number
    """
    digits = [int(dig) for dig in str(number)]
    max_start = len(digits) - consecutive
    return [my_reduce(digits[i:i + consecutive])
            for i in range(max_start + 1)]

def main(verbose=False):
    lines = [line.strip() for line in get_data(8).split('\n')]
    n = "".join(lines)

    return my_max(product_consec_digits(n, 5))

if __name__ == '__main__':
    print euler_timer(8)(main)(verbose=True)
